### ISSUE TITLE
Group info caching

### ISSUE THREAD

**Issue by nickoneill** [5calls/5calls#367]
I think it would be helpful to groups** to show their header (group name, logo, etc) on the top of their call pages as well as the group home page.

Problem is, we only know the group id via the url param and not the group name, so we go fetch that from the server. The roundtrip makes the experience pretty meh.

Let's investigate a way to cache the group info, possibly at `RemoteDataState` for a certain amount of time. This should not just be the current group, but any group we navigate to, so a map from groupid -> group object.

* This could be cached for a certain amount of time, or even just updated in the background every time you visit the group home page.
* We shouldn't refetch the group info on each group call page, but we should fetch it a single time if this is the first time you've visited the group at all and don't have it cached.
* This would also enable us to show the group name in the sidebar rather than "your group" like we do now. That's a much better experience.

** The 5 Calls webapp has the functionality for individuals to form groups (or teams) that allows them to create their own issues (in a separate admin interface) and keep track of the group's call statistics.

**Comment by cdoremus** [comment]
For reference, here's the current `RemoteDataState` structure:
```
export interface RemoteDataState {
 issues: Issue[];
 inactiveIssues: Issue[];
 currentGroup: string;
 groupIssues: Issue[];
 callTotal: number;
 donations: Donations;
 errorMessage: string;
}
```
And here is the current `Group` interface:
```
export interface Group {
  id: string;
  name: string;
  subtitle: string;
  description: string;
  totalCalls: number;
  photoURL: string;
}
```

**Comment by cdoremus** [comment]
Here's my current thinking on this issue:
1. Move `pageGroup` from `GroupPage` local state to `CallState` app state (also add a timestamp field to the `Group` interface).
-- `CallState` is persisted in local storage (`RemoteDataState` is not). 
-- `pageGroup` becomes a `GroupPage` prop.
2. At startup when `startup()` is called, the call state's async action creator is invoked to call `apiServices.getGroup()` (an API call) and set `CallState.pageGroup` by dispatching an action to call state's reducer.
-- the API call would set the `Group` timestamp to the current timestamp.
3. `GroupPageContainer.mapStateToProps()` is used to set the `pageGroup` property using `CallState.pageGroup`.
-- Stale data is checked using `Group.timestamp` (and current timestamp) in `GroupPageContainer.mapStateToProps()`. If the data is stale, the API call is done to refresh `CallState.pageGroup`.
4. A similar data flow will be used to populate group info data in each group call page (`GroupCallPage` using `GroupCallPageContainer`).
-- `GroupCallPage` will also contain a `pageGroup` property.

**TLDR; Summary:** I am moving `GroupPage.pageGroup` to application state (`CallState`) used to set a local prop via `GroupPageContainer` and adding a timestamp to check for stale data. The redux-persist library cache's state in local storage that is used to get group info data instead of the remote API call if it is not stale. This data is used in the group home page and all group call pages.

Please comment.

**Comment by nickoneill** [comment]
Couple questions:

* How does this work with caching information from multiple groups?
* Rather than adding a timestamp to the group object (even though it's not sent with the group object from the server) can we wrap it in a cache container that has the timestamp?

**Comment by cdoremus** [comment]
@nickoneill Thanks for your thoughtful review. Based on your comments and discussions with @schraj on Slack, I will make the following changes to my original design:
* Create a data structure to hold a group (`CachableGroup` below) and a timestamp for use to determine when the group data needs to be refreshed. Note that each group's key in the `CacheableGroup` data structure is the group's id.
```
interface Group {
    id: string;
    name: string;
    description: string;
    /* ... others defined in model.ts */
}

interface CacheableGroup {
    group: { [key: string]: Group };//key is the group id
    timestamp: number;
};

type GroupCache = CacheableGroup[];

// Filled with data from API calls
const groups: GroupCache = [
    {
        group: { craig: { id: 'craig', name: 'Group Craig', description: 'Craig codes' } },
        timestamp: new Date().getTime()
    },
    {
        group: { jeremy: { id: 'jeremy', name: 'Group Jeremy', description: 'Jeremy advises' } },
        timestamp: new Date().getTime()
    },
    {
        group: { nick: { id: 'nick', name: 'Group Nick', description: 'Nick rules!!!' } },
        timestamp: new Date().getTime()
    }
];
```
* Create a cache data structure (`AppCache` below) to hold the array of group data (`GroupCache` above). Additional data could be added to this cache as necessary as separate fields.
```
class AppCache {
  groups: GroupCache;
  constructor(groupCache: GroupCache) {
      this.groups = groupCache;
  }
}
```
* Create a separate Redux store key for the cache (`AppCache`). This will be persisted in local storage using `redux-persist`.

**Comment by nickoneill** [comment]
This sounds great @cdoremus, let's go with this!

**Comment by cdoremus** [comment]
Fixed with PRs #369 and #382
